third_party/nvfuser/csrc/python_frontend/fusion_definition.cpp

```diff
@@ -123,6 +127,41 @@ void FusionDefinition::finalizeDefinition() {
   }
 }
 
+void FusionDefinition::setupSchedule(const at::ArrayRef<c10::IValue>& inputs) {
+  FUSER_PERF_SCOPE("FusionDefinition::setupSchedule");
+  TORCH_CHECK(
+      fusion_id_.has_value(), "FusionDefinition definition does not exist!");
+  auto& scheds = fusionCache()->queryFusionSchedules(fusion_id_.value());
+  auto device = Nvf::getCommonDeviceCUDA(inputs);
+  TORCH_CHECK(device > -1, "Inputs are not all on the same device!");
+  TORCH_CHECK(user_sched_ == nullptr, "Expected User Scheduler to be null!");
+  user_sched_ = fusionCache()->createUserSchedule(scheds, inputs, device);
+
+  // Building a new Fusion container for scheduling with definition such that
+  // the definition's tensor data members refer to the corresponding IR objects
+  // needed for scheduling. A simple copy of the container would mean the data
+  // members that represent tensors would refer to the IR objects in the
+  // original and not the copy needed for scheduling.
+  prev_fusion_ = fusion_;
+  fusion_ = user_sched_->schedule.get();
+  buildFusionIr();
+  fusion_ = prev_fusion_;
+
+  // Manually setting the fusion guard as there is not a good way of using a
+  // guard in a local scope across the schedule function
+  prev_fusion_ = Nvf::FusionGuard::getCurFusion();
+  Nvf::FusionGuard::setCurFusion(user_sched_->schedule.get());
```

Comment:
I don't think it's a good idea to hold on to the FusionGuard between two functions. This implies that we need to hold the lock the entire duration if we want to make it thread-safe and that's not acceptable.

FusionGuard should be lightweight. We should only put on the guard when we are actively modifying it.

Reply:
Same as above.... We talked about this offline in that it is using the FusionGuard methods in an identical way to the local scope based variable where the pointer gotten and set is thread local.